src/zcash/IncrementalMerkleTree.cpp

```diff
@@ -0,0 +1,287 @@
+        if (parent) {
+            root = Hash::combine(*parent, root);
+        } else {
+            root = Hash::combine(root, filler.next());
+        }
+
+        d++;
+    }
+
+    // We may not have parents for ancestor trees, so we fill
+    // the rest in here.
+    while (d < depth) {
+        root = Hash::combine(root, filler.next());
+        d++;
+    }
+
+    return root;
+}
+
+// This constructs an authentication path into the tree in the format that the circuit
+// wants. The caller provides `filler_hashes` to fill in the uncle subtrees.
+template<size_t Depth, typename Hash>
+MerklePath IncrementalMerkleTree<Depth, Hash>::path(std::deque<Hash> filler_hashes) const {
+    if (!left) {
+        throw std::runtime_error("can't create an authentication path for the beginning of the tree");
+    }
+
+    PathFiller<Hash> filler(filler_hashes);
+
+    std::vector<Hash> path;
+    std::vector<bool> index;
+
+    if (right) {
+        index.push_back(true);
+        path.push_back(*left);
+    } else {
+        index.push_back(false);
+        path.push_back(filler.next());
+    }
+
+    size_t d = 1;
+
+    BOOST_FOREACH(const boost::optional<Hash>& parent, parents) {
+        if (parent) {
+            index.push_back(true);
+            path.push_back(*parent);
+        } else {
+            index.push_back(false);
+            path.push_back(filler.next());
+        }
+
+        d++;
+    }
+
+    while (d < Depth) {
+        index.push_back(false);
+        path.push_back(filler.next());
+        d++;
+    }
+
+    std::vector<std::vector<bool>> merkle_path;
+    BOOST_FOREACH(Hash b, path)
+    {
+        std::vector<unsigned char> hashv(b.begin(), b.end());
+        std::vector<bool> tmp_b;
+
+        libzerocash::convertBytesVectorToVector(hashv, tmp_b);
+
+        merkle_path.push_back(tmp_b);
+    }
+
+    std::reverse(merkle_path.begin(), merkle_path.end());
+    std::reverse(index.begin(), index.end());
+
+    return MerklePath(merkle_path, index);
+}
+
+template<size_t Depth, typename Hash>
+std::deque<Hash> IncrementalWitness<Depth, Hash>::uncle_train() const {
+    std::deque<Hash> uncles(filled.begin(), filled.end());
+
+    if (cursor) {
+        uncles.push_back(cursor->root(cursor_depth));
+    }
+
+    return uncles;
+}
+
+template<size_t Depth, typename Hash>
+void IncrementalWitness<Depth, Hash>::append(Hash obj) {
+    if (cursor) {
+        cursor->append(obj);
+
+        if (cursor->is_complete(cursor_depth)) {
+            filled.push_back(cursor->root(cursor_depth));
+            cursor = boost::none;
+        }
+    } else {
+        cursor_depth = tree.next_depth(filled.size());
+
+        if (cursor_depth >= Depth) {
+            throw std::runtime_error("tree is full");
+        }
+
+        if (cursor_depth == 0) {
+            filled.push_back(obj);
+        } else {
+            cursor = IncrementalMerkleTree<Depth, Hash>();
+            cursor->append(obj);
+        }
+    }
+}
```

Comment:
(Thinking out loud) `filled` contains the roots of all subtrees that are complete and to the right of the path to the object we're witnessing. Since they're complete, their roots will never change. `cursor` is the smallest incomplete subtree that's to the right of the path to the object we're witnessing. New objects go into that tree, until we fill it up, and its root becomes the next entry in `filled`. Then we make a new cursor for the next-bigger subtree to the right, and so on.